Kernel/Graphics/VirtIOGPU/VirtIOFrameBufferDevice.cpp

```diff
@@ -43,35 +118,43 @@ int VirtIOFrameBufferDevice::ioctl(FileDescription&, unsigned request, FlatPtr a
         FBResolution resolution;
         if (!copy_from_user(&resolution, user_resolution))
             return -EFAULT;
-        if (!m_gpu->try_to_set_resolution(resolution.width, resolution.height))
+        if (!try_to_set_resolution(resolution.width, resolution.height))
             return -EINVAL;
-        resolution.pitch = m_gpu->framebuffer_pitch();
+        resolution.pitch = pitch();
         if (!copy_to_user(user_resolution, &resolution))
             return -EFAULT;
         return 0;
     }
     case FB_IOCTL_GET_RESOLUTION: {
         auto* user_resolution = (FBResolution*)arg;
         FBResolution resolution;
-        resolution.pitch = m_gpu->framebuffer_pitch();
-        resolution.width = m_gpu->framebuffer_width();
-        resolution.height = m_gpu->framebuffer_height();
+        resolution.pitch = pitch();
+        resolution.width = width();
+        resolution.height = height();
         if (!copy_to_user(user_resolution, &resolution))
             return -EFAULT;
         return 0;
     }
-    case FB_IOCTL_FLUSH_BUFFER: {
-        FBRect user_dirty_rect;
-        if (!copy_from_user(&user_dirty_rect, (FBRect*)arg))
+    case FB_IOCTL_FLUSH_BUFFERS: {
+        FBRects user_dirty_rects;
+        if (!copy_from_user(&user_dirty_rects, (FBRects*)arg))
             return -EFAULT;
-        VirtIOGPURect dirty_rect {
-            .x = user_dirty_rect.x,
-            .y = user_dirty_rect.y,
-            .width = user_dirty_rect.width,
-            .height = user_dirty_rect.height
-        };
-        if (m_are_writes_active)
-            m_gpu->flush_dirty_window(dirty_rect);
+        if (Checked<unsigned>::multiplication_would_overflow(user_dirty_rects.count, sizeof(FBRect)))
+            return -EFAULT;
+        for (unsigned i = 0; i < user_dirty_rects.count; i++) {
+            FBRect user_dirty_rect;
+            if (!copy_from_user(&user_dirty_rect, &user_dirty_rects.rects[i]))
+                return -EFAULT;
+            if (m_are_writes_active) {
```

Comment:
Nit: Maybe this check can be moved outside the entire loop?

Reply:
Then we need to deal with making sure we have enough storage, or do it in chunks. Let's keep it for now, we're going to be improving this thing anyway, there's plenty to do ;-) I am working on some right now.

Reply:
Ah, I meant the m_are_writes_active check, not the copy from user. If m_are_writes_active is false, this whole thing should be a no-op (it also means that the userspace program was writing to the sink region).

Reply:
Agree that this isn't terribly important right now.